### DISCUSSION TITLE
Will AIR support receive postMessage form the webview2?

### DISCUSSION THREAD

**Discussion by sol-oka** [airsdk/Adobe-Runtime-Support#1969] [General]
Because of I have a project need to communicate the web content. In android / iOS webview, I can detect the hash change to provide the app and webview  communication. But in webview2, I can't detect the hash change. And I know the C# and C++ can communicate with webview2 through the postMessage. So, will AIR provide this function?

**Comment by ajwfrost** [comment]
Hi

Currently you can communicate from the webview content to the AS3 content via things like kicking off a location change with parameters (so e.g. we've used something like `window.location="about:blank?some=parameters"` and handling the `locationChanging` event in AS3 to pick that up, work out the parameters, and cancel the navigation.

But we would be looking to introduce a slightly better way of doing this! the challenge of course is to do this in a way that works across all the different browser engines.

Can I check what you're mainly trying to do: do you need to call some AS3 function from JavaScript, or call a JavaScript function from ActionScript, or are you trying to query the DOM or similar?

thanks

**Reply by sol-oka** [reply]
Thanks for your reply.

AIR communicate with the webview2 by "location change with parameters" is not suitable for me. In normal case, I can add hash change detect in the JS. When hash change, I will change the location to special url. Then AIR detect this special url will call preventDefault to prevent the web content change and read the parameters. But in webRTC case, change the location cause the webRTC stop as the same time.

**Reply by ajwfrost** [reply]
Okay thanks for clarifying .. we'll look at how we can set up a better/more direct way for the JavaScript to ActionScript communications...

**Comment by D5CN** [comment]
May be you can try this: https://github.com/D5-Howard/ASHttpd

**Reply by ajwfrost** [reply]
This is a great approach, and would then allow things to be built on top of this to open up more of an API.

FWIW, we were also planning to add support for the postMessage mechanism to the desktop StageWebView objects.. but it's very similar conceptually: just a framework for messages to be sent asynchronously, and a layer can be added to the top to turn the communications into something more meaningful..

thanks

**Comment by itlancer** [comment]
By the way. It's supported starting from AIR 51.
https://airsdk.dev/reference/actionscript/3.0/flash/media/StageWebView.html#postMessage()
https://airsdk.dev/reference/actionscript/3.0/flash/media/StageWebView.html#event:webViewMessage
